
# Week 01 Assignment  
## Data Quality, Evaluation, Scaling, and Encoding

**Student name: Raiyan Ashrafi**   

This is a small assignment that connects topics from Module 1, 2, and 3.  
You must complete it in this Colab notebook.

You will need to use concepts that appeared in the videos:
- Module 1 and 2: basic descriptive statistics, proportions, confusion matrix, accuracy, precision, recall
- Module 3: standardization, min max scaling, nominal vs ordinal, one hot encoding, ordinal encoding, Euclidean and Manhattan distance

Please do not use any extra libraries beyond `pandas`, `numpy`.



---
## 0. Setup and Dataset

We will use a dataset that should have columns given below:

- `user_id`  
- `age`  
- `monthly_income` (numeric)  
- `daily_screen_time_min` (numeric)  
- `daily_app_opens` (numeric)  
- `true_label` and `pred_label` for a binary classification task (0 or 1)  
- `satisfaction_level` (for example: `Low`, `Medium`, `High`)  
- `city_type` (for example: `Urban`, `Suburban`, `Rural`)


In [ ]:
# Cell 1: Imports
import pandas as pd
import numpy as np

In [ ]:
# Cell 2: Load the dataset (Already done for you)
df = pd.read_csv("https://drive.google.com/uc?export=download&id=1OmDDCh4MD1TtvAemnwVDyz5zwCIXJ220")

# Show first few rows
df.head()

user_id  age  monthly_income  daily_screen_time_min  daily_app_opens  \
0        1   43         3734.19                    109               48   
1        2   49         2594.19                    194                7   
2        3   19         3550.47                    146               36   
3        4   19         3821.18                    287               14   
4        5   63         1750.84                     66               46   

   true_label  pred_label satisfaction_level city_type  
0           0           0             Medium  Suburban  
1           0           0                Low     Urban  
2           1           0               High     Rural  
3           1           0               High  Suburban  
4           0           0             Medium  Suburban


### 0.1 Check your dataset

1. Confirm that the dataset loaded correctly.  
2. Check that you have at least these columns:  
   - numeric: `age`, `monthly_income`, `daily_screen_time_min`, `daily_app_opens`  
   - labels: `true_label`, `pred_label`  
   - categorical: `satisfaction_level`, `city_type`  



---
## Part A - Module 1 and 2 Review

In this part you will do simple descriptive statistics and basic classification evaluation.



### Q1. Descriptive statistics on a numeric feature

Choose one numeric column, for example `daily_screen_time_min`.


In [ ]:
# Q1.1: Choose your numeric column here [We already write this ans]
num_col = "daily_screen_time_min"

df[num_col].describe()

count    100.000000
mean     181.890000
std       68.886951
min       60.000000
25%      122.000000
50%      178.000000
75%      243.750000
max      299.000000
Name: daily_screen_time_min, dtype: float64


> **Q1.2 Short answer: [Marks: 05]**  
> Look at the count, mean, min, max, and standard deviation for your chosen column.  
> In 2 to 3 sentences, comment on what you see.  
> For example, does the max look very far from the mean, or does it look quite close?

Write your answer here:

>  The dataset shows the daily screen time in minutes ,showing the average screen time is 181.89 minutes.
>  The minimum value is 60 minutes while the maximum value is 299 minutes.
>  The maximum value is much higher than the mean value which indicates there lies some outliers or it can be caused by right skewed distribution of data points.



### Q2. Proportion of positive class

Use the `true_label` column, where 1 means "positive" and 0 means "negative".


In [ ]:
# Q2.1: Compute proportion of positive class [We already write this ans]
label_col = "true_label"

positive_count = (df[label_col] == 1).sum()
total_count = df.shape[0]
positive_proportion = positive_count / total_count

print("Positive count:", positive_count)
print("Total samples:", total_count)
print("Proportion of positive class:", positive_proportion)

Positive count: 52
Total samples: 100
Proportion of positive class: 0.52



> **Q2.2 Short answer: [5 marks]**  
> In 1 to 2 sentences, explain what this proportion tells you about your dataset.  
> For example, is the dataset balanced between 0 and 1, or is one class much more common?

Write your answer here:

>  The proportion 0.52 tells us that the dataset is almost evenly spilt between two classes.
>  
>  



### Q3. Confusion matrix and basic metrics

For this question, use:
- `true_label` as the actual label  
- `pred_label` as the model prediction


In [ ]:
# Q3.1: Manually compute TP, TN, FP, FN [We already write this ans]
true_col = "true_label"
pred_col = "pred_label"

tp = ((df[true_col] == 1) & (df[pred_col] == 1)).sum()
tn = ((df[true_col] == 0) & (df[pred_col] == 0)).sum()
fp = ((df[true_col] == 0) & (df[pred_col] == 1)).sum()
fn = ((df[true_col] == 1) & (df[pred_col] == 0)).sum()

print("TP:", tp)
print("TN:", tn)
print("FP:", fp)
print("FN:", fn)

TP: 28
TN: 27
FP: 21
FN: 24


In [ ]:
# Q3.2: Compute accuracy, precision, recall [We already write this ans]
accuracy = (tp + tn) / (tp + tn + fp + fn) if (tp + tn + fp + fn) > 0 else 0

precision = tp / (tp + fp) if (tp + fp) > 0 else 0
recall = tp / (tp + fn) if (tp + fn) > 0 else 0

print("Accuracy:", accuracy)
print("Precision:", precision)
print("Recall:", recall)

Accuracy: 0.55
Precision: 0.5714285714285714
Recall: 0.5384615384615384



> **Q3.3 Short answer: [10 marks]**  
> In 3 to 4 sentences, briefly comment on the model using these three metrics.  
> For example, is the model catching most positives (high recall) or being careful when it predicts positive (high precision)?

Write your answer here:

>  The model's accuracy is 0.55, which is better than random guessing.
>  The precision is 0.57 which is higher than recall that is 0.53, which indicates the model is more often right than it is wrong.
>  However, the model's recall is very low which indicates the model is missing lot of positive cases ,so it's predictions are very weak and they can be improved



---
## Part B - Module 3: Scaling and Encoding

Now we will pick a few features and apply scaling and encoding.



### Q4. Standardization and Min max scaling

Use one numeric column, `monthly_income`.


In [ ]:
# Q4.1: Choose the numeric column [2 marks]
col = "daily_screen_time_min"

In [ ]:
# Q4.2: Standardization with z-score [10 marks]
import pandas as pd

url = "https://drive.google.com/uc?export=download&id=1OmDDCh4MD1TtvAemnwVDyz5zwCIXJ220"
df = pd.read_csv(url)

mean = df["daily_screen_time_min"].mean()
std = df["daily_screen_time_min"].std()

df["z_score"] = (df["daily_screen_time_min"] - mean) / std
df.head()

user_id  age  monthly_income  daily_screen_time_min  daily_app_opens  \
0        1   43         3734.19                    109               48   
1        2   49         2594.19                    194                7   
2        3   19         3550.47                    146               36   
3        4   19         3821.18                    287               14   
4        5   63         1750.84                     66               46   

   true_label  pred_label satisfaction_level city_type   z_score  
0           0           0             Medium  Suburban -1.058110  
1           0           0                Low     Urban  0.175795  
2           1           0               High     Rural -0.520999  
3           1           0               High  Suburban  1.525833  
4           0           0             Medium  Suburban -1.682322

In [ ]:
# Q4.3: Min max scaling implementation [10 marks]
import pandas as pd

url = "https://drive.google.com/uc?export=download&id=1OmDDCh4MD1TtvAemnwVDyz5zwCIXJ220"
df = pd.read_csv(url)


min_val = df["daily_screen_time_min"].min()
max_val = df["daily_screen_time_min"].max()

df["minmax_scale"] = (df["daily_screen_time_min"] - min_val) / (max_val - min_val)

df.head()

user_id  age  monthly_income  daily_screen_time_min  daily_app_opens  \
0        1   43         3734.19                    109               48   
1        2   49         2594.19                    194                7   
2        3   19         3550.47                    146               36   
3        4   19         3821.18                    287               14   
4        5   63         1750.84                     66               46   

   true_label  pred_label satisfaction_level city_type  minmax_scale  
0           0           0             Medium  Suburban      0.205021  
1           0           0                Low     Urban      0.560669  
2           1           0               High     Rural      0.359833  
3           1           0               High  Suburban      0.949791  
4           0           0             Medium  Suburban      0.025105


> **Q4.4 Short answer: [3 marks]**  
> Compare the standardized and min max scaled columns in 2 to 3 sentences.  
> Mention what kind of range each one uses and how the numbers look.

Write your answer here:

> The standardized column uses z-score so its value can be positive or negative depending on how far is the value from mean.
>  Printing the head we can see some of the values of z-score is in negative which indicate they are less than mean but the min max column keep all the values between 0 and 1, making them easier to compare without even centering them.
>  



### Q5. One hot and ordinal encoding

We will use:
- `city_type` as a nominal feature  
- `satisfaction_level` as an ordinal feature with order `Low` < `Medium` < `High`  


In [46]:
# Q5.1: One hot encoding for city_type using pandas [10 marks]
import pandas as pd


url = "https://drive.google.com/uc?export=download&id=1OmDDCh4MD1TtvAemnwVDyz5zwCIXJ220"
df = pd.read_csv(url)

df_encoded = pd.get_dummies(df, columns=["city_type"], prefix="city")

df_encoded.head()

user_id  age  monthly_income  daily_screen_time_min  daily_app_opens  \
0        1   43         3734.19                    109               48   
1        2   49         2594.19                    194                7   
2        3   19         3550.47                    146               36   
3        4   19         3821.18                    287               14   
4        5   63         1750.84                     66               46   

   true_label  pred_label satisfaction_level  city_Rural  city_Suburban  \
0           0           0             Medium       False           True   
1           0           0                Low       False          False   
2           1           0               High        True          False   
3           1           0               High       False           True   
4           0           0             Medium       False           True   

   city_Urban  
0       False  
1        True  
2       False  
3       False  
4       False

In [47]:
# Q5.2: Attach one hot encoded columns to df [5 marks]
import pandas as pd


url = "https://drive.google.com/uc?export=download&id=1OmDDCh4MD1TtvAemnwVDyz5zwCIXJ220"
df = pd.read_csv(url)

city_dummies = pd.get_dummies(df["city_type"], prefix="city")

df = pd.concat([df, city_dummies], axis=1)

df.head()

user_id  age  monthly_income  daily_screen_time_min  daily_app_opens  \
0        1   43         3734.19                    109               48   
1        2   49         2594.19                    194                7   
2        3   19         3550.47                    146               36   
3        4   19         3821.18                    287               14   
4        5   63         1750.84                     66               46   

   true_label  pred_label satisfaction_level city_type  city_Rural  \
0           0           0             Medium  Suburban       False   
1           0           0                Low     Urban       False   
2           1           0               High     Rural        True   
3           1           0               High  Suburban       False   
4           0           0             Medium  Suburban       False   

   city_Suburban  city_Urban  
0           True       False  
1          False        True  
2          False       False  
3           True       False  
4           True       False

In [48]:
# Q5.3: Ordinal encoding for satisfaction_level [10 marks]
import pandas as pd

url = "https://drive.google.com/uc?export=download&id=1OmDDCh4MD1TtvAemnwVDyz5zwCIXJ220"
df = pd.read_csv(url)

satisfaction_order = {"Low": 1, "Medium": 2, "High": 3}

df["satisfaction_level_encoded"] = df["satisfaction_level"].map(satisfaction_order)

df.head()

user_id  age  monthly_income  daily_screen_time_min  daily_app_opens  \
0        1   43         3734.19                    109               48   
1        2   49         2594.19                    194                7   
2        3   19         3550.47                    146               36   
3        4   19         3821.18                    287               14   
4        5   63         1750.84                     66               46   

   true_label  pred_label satisfaction_level city_type  \
0           0           0             Medium  Suburban   
1           0           0                Low     Urban   
2           1           0               High     Rural   
3           1           0               High  Suburban   
4           0           0             Medium  Suburban   

   satisfaction_level_encoded  
0                           2  
1                           1  
2                           3  
3                           3  
4                           2


> **Q5.4 Short answer: [5 marks]**  
> In 2 to 3 sentences, explain why one hot encoding is suitable for `city_type`  
> and why ordinal encoding is suitable for `satisfaction_level`.

Write your answer here:

>  Here city_type has no natural ranking order, therefore adding an binary column prevents the model from assuming any ranking between the cities.
>  While has satisfaction_level has an ranking order for an example low, medium and high, so adding a values like 1,2 and 3 preserves the meaning of the order but keeping them in numeric order.
>  



---
## Part C - Module 3: Distances between users

For this small part we will work with vectors based on scaled numeric features.



### Q6. Euclidean and Manhattan distance

Build 2D vectors for user 0 and user 1 using:
- `income_std`  
- `daily_app_opens` (or its min max scaled version if you prefer)


In [49]:
# Q6.1: Build 2D vectors for first two users [We already write this ans]
vec_cols = ["monthly_income", "daily_app_opens"]

v1 = df.loc[0, vec_cols].values
v2 = df.loc[1, vec_cols].values

print("v1:", v1)
print("v2:", v2)

v1: [np.float64(3734.19) np.int64(48)]
v2: [np.float64(2594.19) np.int64(7)]


In [50]:
# Q6.2: Euclidean distance computation [5 marks]
import numpy as np

distance = np.sqrt(np.sum((v1-v2) ** 2))

print("Euclidean distance :", distance)

Euclidean distance : 1140.7370424422975


In [51]:
# Q6.3: Manhattan distance computation [5 marks]
manhattan_distance = np.sum(np.abs(v1-v2))

print("Manhattan distance : ", manhattan_distance)

Manhattan distance :  1181.0



> **Q6.4 Short answer: [5 marks]**  
> Which one is larger in your result, Euclidean or Manhattan distance  
> and why does that usually happen based on their formulas?

Write your answer here:

>  The manhattan value is larger than the euclidean distance because in manhattan distance we are taking the absolute value while in euclidean distance we are taking the square root of the sum of the squared distance differences.
>  For that reason usually the manhattan distance is larger than the euclidean distance.
>  



---
## Final Reflection [10 marks]

> In 4 to 6 sentences, describe how the three modules connect in this assignment.  
> Mention:
> - One idea from Module 1 or 2 that you used  
> - One idea from Module 3 that you used  
> - How these ideas together help you understand a dataset more deeply

Write your reflection here:

>  The ideas I've used from module 1 and 2 are descriptive statistics and proportions, to understand the central tendency, spread, and balance of the dataset.
>  From module 3 I've used standardization, min max scaling to understand the distance between data points. Combining all these it has helped me to understand the data points for model training and analyzing.
>  In short the forumlas have helped me to understand the relation and pattern between all the columns of the data set.



## End of Assignment

Before submitting:
- Run all cells from top to bottom.  
- Check that all answer sections are filled.  
- Instruction video অনুযায়ী আমাদের দেয়া Colab ফাইলটি থেকে প্রথম একটি Save copy in drive করে নিবা। এরপর Google colab এর মধ্যে কোডগুলো করবে এবং সেই ফাইলটি ‘Anyone with the link’ & ‘View’ Access দিয়ে ফাইলটির Shareble Link টি সাবমিট করবে।
